packages/graphql-plugin-auth/src/JWKSPlugin.ts

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) "Neo4j"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import jsonwebtoken from "jsonwebtoken";
+import JwksRsa, { JwksClient } from "jwks-rsa";
+import Debug from "debug";
+import { DEBUG_PREFIX } from "./constants";
+
+const debug = Debug(DEBUG_PREFIX);
+
+function verifyJWKS<T = any>({ client, token }: { client: JwksClient; token: string }): Promise<T> {
+    const getKey: jsonwebtoken.GetPublicKeyOrSecret = (header, callback) => {
+        const kid: string = header.kid || "";
+
+        client.getSigningKey(kid, (err, key) => {
+            const signingKey = key?.getPublicKey();
+            callback(err, signingKey);
+        });
+    };
+
+    // Returns a Promise with verification result or error
+    return new Promise((resolve, reject) => {
+        jsonwebtoken.verify(
+            token,
+            getKey,
+            {
+                algorithms: ["HS256", "RS256"],
+            },
+            (err, decoded) => {
+                if (err) {
+                    reject(err);
+                } else {
+                    resolve(decoded as unknown as T);
+                }
+            }
+        );
+    });
+}
+
+export interface JWKSPluginInput {
+    jwksEndpoint: string;
+    rolesPath?: string;
+    clientOptions?: JwksRsa.Options;
+}
+
+class JWKSPlugin {
```

Comment:
I think we have some kind of discussion, but wouldn't it make sense to use an interface for these plugins?

```ts
interface Neo4jGraphQLPAuthPluginInterface {
    rolesPath?: string
    decode<T>(token: string): Promise<T | undefined>;
}
```

```suggestion
class JWKSPlugin implements Neo4jGraphQLPAuthPluginInterface {
```

This way we ensure consistency over the different plugins. In the future this interface should be shared between libraries, but for now should be enough

Reply:
Because `@neo4j/graphql` imports `@neo4j/graphql-plugin-auth` you cant do this otherwise you get a circ dependency.

Reply:
Oh, no, I'm talking about creating an interface **inside** `graphql-plugin-auth`

It may be a small code duplication there, but I think is worth it as we have to plugins that should follow the same interface